DM-46129: Make collections.query_info a single HTTP call

dhirving opened the pull request:
Previously `RemoteButlerCollections.query_info` would make an HTTP call for each resolved collection.  It now does everything in a single call to the server instead.

`RemoteButlerCollections.get_info` and other collection methods in `RemoteButlerRegistry` have also been updated to re-use the new endpoint.

Also fixed an issue where the `summary_datasets` parameter to `DirectButlerCollections.query_info` was not doing anything.

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

dhirving committed:
Make collections.query_info a single HTTP call

Previously collections.query_info would make an HTTP call for each resolved collection.  It now does everything in a single call to the server instead.

dhirving committed:
Make get_info use the same endpoint as query_info

There is no reason to add a duplicate API for this, because get_info is identical to calling query_info with a single collection name.

dhirving committed:
Break RegistryDefaults circular dependency

The various Butler sub-objects (registry, collections) need to access the registry defaults.  Added a small helper class so that they don't need a reference to RemoteButler or each other in order to fetch the defaults.

dhirving committed:
Forward RemoteButlerRegistry to ButlerCollections

Now that there is an independent implementation of ButlerCollections for RemoteButler, forward the equivalent Registry methods to it to reduce duplication.

dhirving committed:
Mark deprecated REST endpoints

andy-slac commented:
DM-45993 was merged, please rebase and add forwarding of new `query_info` parameters to remote server.

dhirving committed:
Handle collection query optimization parameters

Update RemoteButlerCollections to pass the new optimization parameters added in DM-45993 to the server.

dhirving committed:
Fix summary_datasets being ignored

When attempting to add a test for the summary_datasets parameter to ButlerCollectionInfo for RemoteButler, it turned out that it does not doing anything in the DirectButler version.

This was occurring because a caching context causes this parameter to be ignored, and the cache was always enabled in query_info previously.  This cache does not have a benefit for the new implementation.

andy-slac commented on python/lsst/daf/butler/_dataset_type.py:
```diff
@@ -796,3 +796,21 @@ def _unpickle_via_factory(factory: Callable, args: Any, kwargs: Any) -> DatasetT
     return factory(*args, **kwargs)
+
+
+def get_dataset_type_name(datasetTypeOrName: DatasetType | str) -> str:
```
Maybe make it a classmethod of `DatasetType` to avoid extra imports?

andy-slac commented on python/lsst/daf/butler/registry/datasets/byDimensions/_manager.py:
```diff
@@ -12,6 +12,7 @@
 import sqlalchemy
 
 from ...._dataset_ref import DatasetId, DatasetIdGenEnum, DatasetRef, DatasetType
+from ...._dataset_type import get_dataset_type_name
```
`DatasetType` above should also be coming from `_dataset_type`.

andy-slac commented on python/lsst/daf/butler/_butler_collections.py:
```diff
@@ -280,7 +283,7 @@ def query_info(
         include_summary: bool = False,
         include_doc: bool = False,
-        summary_datasets: Iterable[DatasetType] | None = None,
+        summary_datasets: Iterable[DatasetType] | Iterable[str] | None = None,
```
Should we allow `Iterable[str]` only to avoid all complications? I think that clients of this method that actually pass `summary_datasets` should already have a list of names for other purposes. I guess this also means that datasets manager has to accept `Iterable[str]` instead of `Iterable[DatasetType]`?

dhirving replied:
I was thinking the same thing, but the convention in the rest of the new query system is that dataset types can be specified as either the name or `DatasetType` instance so it makes sense to follow that.  It's also common for this parameter to come from the result of a queryDatasetTypes call.

andy-slac commented on python/lsst/daf/butler/_dataset_type.py:
```diff
@@ -796,3 +796,21 @@ def _unpickle_via_factory(factory: Callable, args: Any, kwargs: Any) -> DatasetT
+    datasetTypeOrName : `DatasetType` | `str`
+        A DatasetType, or the name of a DatasetType. This union is a common
+        parameter in many `Butler` methods.
+    """
```
Docstring also needs `Returns`.

andy-slac reviewed (approved):
Looks great. Thanks for adding a unit test for new `query_info` parameters! Maybe we should try to simplify things by passing dataset type names instead of dataset types, but that is up to you.

dhirving committed:
Fix minor review comments
